src/librustc_typeck/check/coercion.rs

```diff
@@ -763,7 +763,7 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
 
         let (adjustments, _) = self.register_infer_ok_obligations(ok);
         self.apply_adjustments(expr, adjustments);
-        Ok(target)
+        Ok(self.resolve_type_vars_with_obligations(target))
```

Comment:
Is there a reason for this? Just to improve debug output, or something else?

Reply:
It's used to avoid repeatedly doing this in the callers.

Reply:
So you mean that the callers rely on type variables having been resolved in the return value?

Reply:
Yeah, in the changed binop code.

Reply:
Then I think we should add some comment that this is important, and also comment the interface of the coercion function. I'm not all that keen on having such a guarantee, though -- it's a subtle link between code imo -- but if you can at least document who cares about it, it would be helpful.

Reply:
Well then I'll just throw away most of the changes, I guess... I thought this API was much nicer but I can just as well have two extra lines in `op.rs` instead.